release_notes/model.py

```diff
@@ -0,0 +1,228 @@
+    def get_identifier(self) -> str:
+        ''' returns a human-readable identifier which can be used e.g. for duplicate checking.
+        does not include line breaks or spaces
+        '''
+        return f'[{self.category}<{self.target_group}>]{self.note_message}' \
+            .lower().strip().replace(' ', '').replace('\n', '')
+
+    def has_content(self) -> bool:
+        ''' checks if there is any content in the source note block
+        '''
+        if self.note_message.strip().lower() == 'none':
+            return False
+        return all(z and z.strip() for z in (self.category, self.target_group, self.note_message))
+
+    def __hash__(self):
+        return hash(self.get_identifier())
+
+    def __eq__(self, other):
+        if isinstance(other, ReleaseNote):
+            return self.__eq__(other.source_block)
+        if isinstance(other, SourceBlock):
+            return hash(other) == hash(self)
+        return False
+
+
+pattern = re.compile(r'\x60{3}(?P<category>\w+)\s+(?P<target_group>\w+)\n(?P<note>.+?)\n\x60{3}',
+                     flags=re.DOTALL | re.IGNORECASE | re.MULTILINE)
+
+
+def iter_source_blocks(content: str) -> typing.Generator[SourceBlock, None, None]:
+    ''' Searches for code blocks in release note notation and returns all found.
+    Only valid note blocks are returned, which means that the format has been followed.
+    However, it does not check if the category / group exists.
+
+    :param content: the content to look for release notes in
+    :return: a list of valid note blocks
+    '''
+    for res in pattern.finditer(content.replace('\r\n', '\n')):
+        try:
+            block = SourceBlock(category=res.group('category'),
+                                target_group=res.group('target_group'),
+                                note_message=res.group('note'))
+            if block.has_content():
+                yield block
+        except IndexError:
+            # group not found, ignore
+            continue
+
+
+@dataclasses.dataclass
+class ReleaseNote:
+    source_commit: Commit  # the commit where the release notes were initially gathered from
+    source_block: SourceBlock
+
+    raw_body: str  # the raw body of the commit / pull request
+    author: typing.Optional[Author]  # the author of the commit / pull request
+    reference: Reference
+
+    source_component: gci.componentmodel.Component = dataclasses.field(compare=False)
+    is_current_repo: bool
+    from_same_github_instance: bool
+
+    def __hash__(self):
+        return hash(self.source_block.get_identifier())
+
+    def __eq__(self, other) -> bool:
+        return self.source_block.__eq__(other)
+
+    def get_author_str(self) -> str:
+        return str(self.author)
+
+    def get_reference_str(self) -> str:
+        return f'{self.reference.type.identifier}{self.reference.get_identifier()}'
+
+    def to_block_str(self) -> str:
+        return '```{category} {group} {src_repo} {ref} {author}\n{message}\n```'.format(
```

Comment:
preferably use format-string-literals: https://docs.python.org/3/reference/lexical_analysis.html#f-strings

Reply:
I think using format instead of f-strings helps with readability, wdyt?

```py
    def to_block_str(self) -> str:
        return '```{category} {group} {src_repo} {ref} {author}\n{message}\n```'.format(
            category=self.source_block.category,
            group=self.source_block.target_group,
            src_repo=self.source_component.name,
            ref=self.get_reference_str(),
            author=self.author.username or self.author.display_name.replace(' ', '-'),
            message=self.source_block.note_message
        )
```

vs.

```py
    def to_block_str(self) -> str:
        return f'```{self.source_block.category} {self.source_block.target_group} {self.source_component.name} ' \
               f'{self.get_reference_str()} {self.author.username or self.author.display_name.replace(" ", "-")}\n' \
               f'{self.source_block.note_message}\n```'
```

or maybe a mix of both?

```py
    def to_block_str(self) -> str:
        category = self.source_block.category
        group = self.source_block.target_group
        src_repo = self.source_component.name
        ref = self.get_reference_str()
        author = self.author.username or self.author.display_name.replace(' ', '-')
        message = self.source_block.note_message

        return f'```{category} {group} {src_repo} {ref} {author}\n{message}\n```'
```